lambda/matching.ml

```diff
@@ -2819,48 +2880,92 @@ let combine_constructor loc arg pat_env cstr partial ctx def
           mk_failaction_pos partial constrs ctx def
       in
       let descr_lambda_list = fails @ descr_lambda_list in
-      let consts, nonconsts =
-        split_cases (List.map tag_lambda descr_lambda_list) in
       let lambda1 =
         match (fail_opt, same_actions descr_lambda_list) with
         | None, Some act -> act (* Identical actions, no failure *)
         | _ -> (
+            let cases = split_cases (List.map tag_lambda descr_lambda_list) in
             match
-              (cstr.cstr_consts, cstr.cstr_nonconsts, consts, nonconsts)
+              (cstr.cstr_consts, cstr.cstr_nonconsts, cases)
             with
-            | 1, 1, [ (0, act1) ], [ (0, act2) ] ->
+            | 1, 1, { consts = [ (0, act1) ]; nonconsts = [ (0, act2) ] }
                 (* Typically, match on lists, will avoid isint primitive in that
               case *)
+            | 1, 0, { consts = [ (0, act1) ]; any_nonconst = Some act2 } ->
                 Lifthenelse (arg, act2, act1)
-            | n, 0, _, [] ->
+            | n, 0, { any_const; nonconsts = []; any_nonconst = None } ->
                 (* The type defines constant constructors only *)
-                call_switcher loc fail_opt arg 0 (n - 1) consts
-            | n, _, _, _ -> (
-                let act0 =
-                  (* = Some act when all non-const constructors match to act *)
-                  match (fail_opt, nonconsts) with
-                  | Some a, [] -> Some a
-                  | Some _, _ ->
-                      if List.length nonconsts = cstr.cstr_nonconsts then
-                        same_actions nonconsts
+               begin match any_const with
+               | Some act -> act
+               | None ->
+                  call_switcher loc fail_opt arg 0 (n - 1) cases.consts
+               end
```

Comment:
nitpick: it's weird that we match on `any_const` in the record pattern, but access `cases.consts` by projection. I would use the same approach for both.